src/box/box.cc

```diff
@@ -1805,6 +1805,214 @@ box_wait_quorum(uint32_t lead_id, int64_t target_lsn, int quorum,
+	struct fiber *waiter;
+	/** Collected vclock. */
+	struct vclock *vclock;
+	/** The request deadline. */
+	double deadline;
+	/** Whether the request is timed out. */
+	bool timed_out;
+};
+
+/**
+ * A trigger executed on each ack to collect up to date remote node vclocks.
+ * When an ack comes with requested sync for some replica id, ack vclock is
+ * accounted.
+ */
+static int
+check_sync_on_ack(struct trigger *trigger, void *event)
+{
+	struct replication_ack *ack = (struct replication_ack *)event;
+	struct sync_trigger_data *data =
+		(struct sync_trigger_data *)trigger->data;
+	uint32_t id = ack->source;
+	uint64_t sync = data->syncs[id];
+	if (!bit_test(&data->collected_vclock_map, id) && sync > 0 &&
+	    ack->sync >= sync) {
+		vclock_max_ignore0(data->vclock, ack->vclock);
+		bit_set(&data->collected_vclock_map, id);
+		fiber_wakeup(data->waiter);
+	}
+	return 0;
+}
+
+/** A trigger querying relay's next sync value once it becomes operational. */
+static int
+relay_get_sync_on_start(struct trigger *trigger, void *event)
+{
+	struct replica *replica = (struct replica *)event;
+	if (replica->anon)
+		return 0;
+	struct relay *relay = replica->relay;
+	struct sync_trigger_data *data =
+		(struct sync_trigger_data *)trigger->data;
+	uint32_t id = replica->id;
+	try {
+		if (!bit_test(&data->collected_vclock_map, id)) {
+			data->syncs[id] = relay_next_sync(relay,
+							  data->deadline);
+		}
+	} catch (TimedOut *) {
+		data->timed_out = true;
+		fiber_wakeup(data->waiter);
+	}
+	return 0;
+}
+
+/** Find the minimal vclock which has all the data confirmed on a quorum. */
+static int
+box_collect_confirmed_vclock(struct vclock *confirmed_vclock, double deadline)
+{
+	/*
+	 * How many vclocks we should see to be sure that at least one of them
+	 * contains all data present on any real quorum.
+	 */
+	int vclock_count = MAX(1, replicaset.registered_count) -
+			   replication_synchro_quorum + 1;
+	/*
+	 * We should check the vclock on self plus vclock_count - 1 remote
+	 * instances.
+	 */
+	vclock_copy(confirmed_vclock, &replicaset.vclock);
+	if (vclock_count <= 1)
+		return 0;
+
+	struct sync_trigger_data data = {
+		.syncs = {0},
+		.collected_vclock_map = 0,
+		.waiter = fiber(),
+		.vclock = confirmed_vclock,
+		.deadline = deadline,
+		.timed_out = false,
+	};
+	struct trigger on_relay_thread_start;
+	trigger_create(&on_relay_thread_start, relay_get_sync_on_start, &data,
+		       NULL);
+	trigger_add(&replicaset.on_relay_thread_start, &on_relay_thread_start);
+	struct trigger on_ack;
+	trigger_create(&on_ack, check_sync_on_ack, &data, NULL);
+	trigger_add(&replicaset.on_ack, &on_ack);
+	replicaset_foreach(replica) {
+		if (relay_get_state(replica->relay) != RELAY_FOLLOW ||
+		    replica->anon) {
+			continue;
+		}
+		/* Might be already filled by on_relay_thread_start trigger. */
+		if (data.syncs[replica->id] == 0) {
+			data.syncs[replica->id] =
+				relay_next_sync(replica->relay, deadline);
```

Comment:
Since all `relay_next_sync()` calls go to the same endpoint (tx), their returns are ordered.
Now, while an old relay can still write a message, it's state is still "FOLLOW", because it's changed to "STOPPED" only when the relay thread is joined.
So either the old relay fails with no response at all or its message will be first in queue. The order will be correct.

Have I answered your question?

Reply:
Probably :) Still I wouldn't count on that, but can keep it like now too.